webapp/channels/src/components/admin_console/boolean_setting.tsx

```diff
@@ -74,83 +74,90 @@ type Props = {
     onChange: (id: string, value: boolean) => void;
     trueText?: React.ReactNode;
     falseText?: React.ReactNode;
-    disabled: boolean;
+    disabled: boolean|undefined;
     setByEnv: boolean;
     disabledText?: React.ReactNode;
     helpText: React.ReactNode;
 }
 
-export default class BooleanSetting extends React.PureComponent<Props> {
-    public static defaultProps = {
-        trueText: (
-            <FormattedMessage
-                id='admin.true'
-                defaultMessage='True'
-            />
-        ),
-        falseText: (
-            <FormattedMessage
-                id='admin.false'
-                defaultMessage='False'
-            />
-        ),
-        disabled: false,
+const BooleanSetting = ({
+    id,
+    label,
+    value,
+    onChange,
+    trueText = (
+        <FormattedMessage
+            id='admin.true'
+            defaultMessage='True'
+        />
+    ),
+    falseText = (
+        <FormattedMessage
+            id='admin.false'
+            defaultMessage='False'
+        />
+    ),
+    disabled = false,
+    setByEnv,
+    disabledText,
+    helpText,
+}: Props) => {
+    const [helptext, setHelptext] = useState(helpText);
+    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+        onChange(id, e.target.value === 'true');
     };
 
-    private handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
-        this.props.onChange(this.props.id, e.target.value === 'true');
-    };
-
-    public render() {
-        let helpText;
-        if (this.props.disabled && this.props.disabledText) {
-            helpText = (
+    useEffect(() => {
+        if (disabled && disabledText) {
+            setHelptext((
                 <div>
                     <span className='admin-console__disabled-text'>
-                        {this.props.disabledText}
+                        {disabledText}
                     </span>
-                    {this.props.helpText}
+                    {helpText}
                 </div>
-            );
+            ));
         } else {
-            helpText = this.props.helpText;
+            setHelptext(helpText);
         }
+    }, [helpText, disabled, disabledText]);
```

Comment:
Why storing it in a state and using and effect?

The downside I see for this is that whenever you mount the component, this effect will run, and will set the state, creating a new render. This may even flash the component as enabled when you mount it disabled. Also, whenever some of the dependencies changes (like `helpText`) will generate one re-render (because the prop change), the effect will run, and create another re-render (because the element changes).

I understand the reason why you are doing it this way is to avoid passing new references on every re-render to the `helpText` prop in `Setting`. If that is the case, you would do better with `useMemo`. `useMemo` is executed during the render, so it will update the value in place without creating new re-renders.